### ISSUE TITLE
rustc 1.32.0 produces faulty wasm code

### ISSUE THREAD

**Issue by hoodie** [rust-lang/rust#58548]
Hi there,
rustc 1.32.0 produces different (wrong) wasm code then 1.31.0.
This [example](https://github.com/hoodie/rustwasmbug1320) will yield different results when compiled with rustc 1.31.0 and 1.32.0.

`let x_offset = if yi % 2 == 1 { 23f64 } else { 0f64 };` inside a loop will be repeatedly the same value unless you use `yi` again for instance inside a print. (see [line15](https://github.com/hoodie/rustwasmbug1320/blob/master/src/lib.rs#L15)). This has a very visible effect in a game @payload and I built.

Here you can see the version built with 1.31 and with 1.32
![awsm-gorillas-rustc-1 31 0](https://user-images.githubusercontent.com/260370/52918960-80c4b300-32fd-11e9-8f72-bfb1c1a536af.png)
![awsm-gorillas-rustc-1 32 0](https://user-images.githubusercontent.com/260370/52918964-84583a00-32fd-11e9-808d-2a997308f74c.png)

Please let me know if you have questions about the reproducers.

Thank you

**Comment by nikic** [comment]
I'm pretty sure this is a duplicate of #57152. Can you please check whether the issue is fixed in nightly?

**Comment by jonas-schievink** [comment]
It still outputs "generated code is broken" on the current nightly (`rustc 1.34.0-nightly (eac09088e 2019-02-15)`)

**Comment by hoodie** [comment]
@jonas-schievink you beat me to it, I was just gonna confirm :grin:

**Comment by aturon** [comment]
cc @alexcrichton

**Comment by pnkfelix** [comment]
triage: P-high. Removing I-nominated tag.

**Comment by michaelwoerister** [comment]
cc @alexcrichton

**Comment by alexcrichton** [comment]
Thanks for the report @hoodie, and also thanks so much for for the reduction here! Made it pretty easy to go a bit further.

I've reported this as an upstream LLVM bug to https://bugs.llvm.org/show_bug.cgi?id=40805

**Comment by hoodie** [comment]
just for the fun of it I uploaded both versions of the game
built with: 1.31 http://naubino.de/awsm-gorillas/
built with: 1.32 http://naubino.de/awsm-gorillas/broken/

**Comment by pnkfelix** [comment]
Visiting for triage. @nikic, are we going to cherry-pick that LLVM fix into our local rust-lang/llvm-project?

**Comment by nikic** [comment]
@pnkfelix The fix has been backported to LLVM 8.0 and is already part of our fork. We're just waiting on the update in #58408 to land.

**Comment by nikic** [comment]
The LLVM update has landed in the meantime. @hoodie Could you please double check that the issue is indeed fixed?

**Comment by hoodie** [comment]
it's a bit tricky to tell, because with the current nightly it no longer compiles.
```
error[E0034]: multiple applicable items in scope
   --> /Users/hendrik/.cargo/registry/src/github.com-1ecc6299db9ec823/downcast-0.9.2/src/lib.rs:120:38
    |
120 |     fn is_type(&self) -> bool { self.type_id() == TypeId::of::<T>() }
    |                                      ^^^^^^^ multiple `type_id` found
    |
note: candidate #1 is defined in the trait `Any`
   --> /Users/hendrik/.cargo/registry/src/github.com-1ecc6299db9ec823/downcast-0.9.2/src/lib.rs:29:5
    |
29  |     fn type_id(&self) -> TypeId { TypeId::of::<Self>() }
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: to disambiguate the method call, write `Any::type_id(&self)` instead
note: candidate #2 is defined in the trait `std::any::Any`
    = help: to disambiguate the method call, write `std::any::Any::type_id(&self)` instead
```

was there a way to override the version of dependencies of dependencies with cargo?

**Comment by pnkfelix** [comment]
@hoodie Does `cargo update` not get you the latest version of `downcast` ? Is some dependency asking specifically for `0.9`?

**Comment by pnkfelix** [comment]
In any case I believe you should be able to override arbitrary dependencies (or dependencies of dependencies, etc) via the methods specified at https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#overriding-dependencies

**Comment by nikomatsakis** [comment]
@nikic @hoodie any updates on the issue?

**Comment by hoodie** [comment]
sorry, I didn't manage to build it again, I'll try to put aside some time this weekend

**Comment by nikic** [comment]
Reasonably sure this has been fixed, so closing this.

**Comment by hoodie** [comment]
@nikic sorry, I wasn't able to build the same codebase with the newer compiler again
